BxlAsLib.txt

```diff
@@ -0,0 +1,377 @@
+        }
+        public static Data Deserialize(Reader reader)
+        {
+            Data data;
+            while (!int pos = reader.ReadPos()) != -1) {
+                switch (pos) {
+                    case 1: { data.Path = reader.ReadPath(); break; }
+                    case 2: { data.File = reader.ReadFile(); break; }
+                }
+            }
+            return data;
+        }
+    }
+
+    - Pros: Can use custom serialization/deserialization.
+    - Cons: 
+      - Manually maintained the position.
+      - An intern working on XLG++ may choose protobuf for versioning.
+
+- Excerpt of ProcessData: basically it reflects Process in BuildXL.Pips
+
+  class ProcessData 
+  {
+      /* TBD: We may want to add positional attrbute for each property, like protobuf-net. */
+
+      public FileArtifact Executable { get; set; }
+      public PipData Arguments { get; set; }
+      public AbsolutePath WorkingDirectory { get; set; }
+      public IEnumerable<EnvironmentVariable> EnvironmentVariables { get; set; }
+      public ProcessOptions ProcessOptions { get; set; }
+      public IEnumerable<FileArtifact> Dependencies { get; set; }
+      public IEnumerable<DirectoryArtifact> DirectoryDependencies { get; set; }
+      public IEnumerable<FileArtifactWithAttributes> FileOutputs { get; set; }
+      public IEnumerable<DirectoryArtifact> DirectoryOutputs { get; set; }
+      ...
+  }       
+
+Graph fragment interactions
+---------------------------
+- We want graph fragments to be able to consume values from other frontends and produce values
+  that can be consumed by other frontends.
+- Even when stitching multiple graphs, we want to estabilish dependency relation between fragments.
+- API:
+  - To expose outputs to other frontends or other fragments:
+
+    interface IGraphFragment
+    {
+        ...
+        /* The implementation may verify that this fragment indeed produces the output file. */
+        AddOutputFileValue(FullSymbol id, FileArtifact outputFile);
+        AddOutputDirectoryValue(FullSymbol id, DirectoryArtifact outputDirectory);
+    }
+
+  - To consume file/directory outputs from other frontends or other fragments:
+    class SymbolicDependency
+    {
+        /* Module name can be name of another graph fragment. */
+        public string Module { get; set; }
+        public FullSymbol Id { get; set; }
+    }
+
+    class ProcessData
+    {
+        ...
+        /**
+         * Upon deserialization, the entry in these symbolic dependencies are assumed
+         * to exist (by querying FrontEndHost).
+         * Given (module: M, id: x.y.z) as a symbolic dependency, one should not evaluate
+         * "import('M').x.y.z" because the topological order will be violated.
+         */
+        public IEnumerable<SymbolicDependency> SymbolicDependencies { get; set; }
+        public IEnumerable<SymbolicDependency> SymbolicDirectoryDependencies { get; set; }

```

Comment:
why do you need to differentiate between files and directories here?  It seems that just declaring symbolic dependencies should be enough